kernel/cClosure.ml

```diff
@@ -870,6 +882,58 @@ let drop_parameters depth n argstk =
   (* we know that n < stack_args_size(argstk) (if well-typed term) *)
   anomaly (Pp.str "ill-typed term: found a match on a partially applied constructor.")
 
+(* Iota-reduction: feed the arguments of the constructor to the branch *)
+let get_branch infos depth ci u pms (ind, c) br e args =
+  let i = c - 1 in
+  let args = drop_parameters depth ci.ci_npar args in
+  let (_nas, br) = br.(i) in
+  if Int.equal ci.ci_cstr_ndecls.(i) ci.ci_cstr_nargs.(i) then
+    (* No let-bindings in the constructor, we don't have to fetch the
+      environment to know the value of the branch. *)
+    let rec push e stk = match stk with
+    | [] -> e
+    | Zapp v :: stk -> push (subs_consv v e) stk
+    | (Zshift _ | ZcaseT _ | Zproj _ | Zfix _ | Zupdate _ | Zprimitive _) :: _ ->
+      assert false
+    in
+    let e = push e args in
+    (br, e)
+  else
+    (* The constructor contains let-bindings, but they are not physically
+        present in the match, so we fetch them in the environment. *)
```

Comment:
I understand what you try to do, but when I write
```coq
match t with
| K x y => b
```
The body `b` can only mention `x` and `y`, even if `K : forall x, let w := .. in forall y, I`.

Is my understanding simplistic?

Reply:
I guess I see what I missed, you need them for the type of y :-/

Reply:
I got confused again, here we are reducing, no types.

Reply:
The context of the branch is the whole constructor context, including let-bindings. It might be the case that most user-facing functions prevent you to depend on the let-value, but it's definitely not enforce neither by the typing rules nor by internal functions. And indeed, there were cases where conversion depends on the let-value that were discovered in a previous state of the code that treated those let-bindings as variables. I added some tests for that in other PRs.

Reply:
Actually, the let binders are available in the user-facing syntax.
```
Inductive foo := Foo (x := 0) : foo.

Definition bar (e : foo) :=
match e with Foo x => x end.

Eval compute in bar Foo.
```